File: 0826-soup-servings/0826-soup-servings.py
class Solution:
    def soupServings (self, n : int) -> float:
        def dfs (A , B) :
            A = int (A)
            B = int (B)
            if A <= 0 and B > 0 :
                return 1.0
            if A <= 0 and B <= 0 :
                return 0.5
            if A > 0 and B <= 0 :
                return 0.0
            if dp [A] [B] != -1 :
                return dp [A] [B]
            ans = 0.25 * (
                dfs (A - 4, B) +
                dfs (A - 3, B - 1) +
                dfs (A - 2, B - 2) +
                dfs (A - 1, B - 3)
            )
            dp [A] [B] = ans
            return ans
        dp = [[-1] * 201 for _ in range (201)]
        if n > 5000 :
            return 1.0
        for i in range (201) :
            for j in range (201) :
                dp [i] [j] = -1
        N = int (ceil (n / 25.0))
        return dfs (N, N)

File: 0826-soup-servings/0826-soup-servings.java
class Solution {
    double myCache [] [] = new double [179] [179];
    private double soupServingsHelper (int a, int b) {
        if (a <= 0 && b <= 0) 
            return 0.5;
        else if (a <= 0) 
            return 1;
        else if (b <= 0) 
            return 0;
        else if (myCache [a] [b] != 0) 
            return myCache [a] [b]; 
        else {
            double output1 = soupServingsHelper (a - 4, b);
            double output2 = soupServingsHelper (a - 3, b - 1);
            double output3 = soupServingsHelper (a - 2, b - 2);
            double output4 = soupServingsHelper (a - 1, b - 3);
            if (a >= 4)
            myCache [a - 4] [b] = output1;
            if (a >= 3 && b >= 1) 
            myCache [a - 3] [b - 1] = output2;
            if (a >= 2 && b >= 2) 
            myCache [a - 2] [b - 2] = output3;
            if (a >= 1 && b >= 3)
            myCache [a - 1] [b - 3] = output4;
            return 0.25 * (output1 + output2 + output3 + output4);
        }
    }   
    public double soupServings (int n) {
        if (n >= 4451)
            return 1;
        n = (n + 24) / 25;
        return soupServingsHelper (n, n);
    }
}